typed-racket-lib/typed-racket/static-contracts/optimize.rkt

```diff
@@ -119,6 +119,11 @@
     [(none/sc:) any/sc]
     [(or/sc: (? flat-terminal-kind?) ...) any/sc]
     [(? flat-terminal-kind?) any/sc]
+    [(syntax/sc: (? recursive-sc?))
+     ;;bg; _temporary_ case to allow contracts from the `Syntax` type.
```

Comment:
Do we need the `(? recursive-sc?)`? Doesn't anything there _have_ to be something we can check with a flat-contract and so we can optimize any `syntax/sc:` to `any/sc`?

Reply:
I want to keep the `(? recursive-sc?)`.

If we optimize any `syntax/sc:` to `any/sc`, then a type like `(Syntaxof (-> Integer Integer))` would get optimized to `any/sc` instead of raising an exception.

- - -

(while I'm in the neighborhood...)

Why `recursive-sc?` and not anything stronger? Checking `recursive-sc?` is a coarse way of looking for the contract I actually care about (the static contract from `Syntax`), but:
- I don't think I can match against the actual recursive contract
- it seems unlikely that someone has code with a non-flat recursive type inside a `Syntaxof` (even less likely than having a function inside)
- the only other strategy I can think of to fix the issue here is changing `type->contract` (which can definitely recognize the `Syntax` type), but then `type->contract` would need to know whether the current position is one that's safe to optimize